# [Introduction to Data Science: A Comp-Math-Stat Approach](http://datascience-intro.github.io/1MS041-2021/)    
## 1MS041, 2021 
&copy;2021 Raazesh Sainudiin, Benny Avelin. [Attribution 4.0 International     (CC BY 4.0)](https://creativecommons.org/licenses/by/4.0/)

# 01. BASH Unix Shell 


1. Dropping into BASH (Unix Shell) and using basic Shell commands
    * `pwd` --- print working directory
    * `ls` --- list files in current working directory
    * `mkdir` --- make directory
    * `cd` --- change directory
    * `man ls` --- manual pages for any command
2. Grabbing files from the internet using `curl`

In [1]:
def showURL(url, ht=500):
    """Return an IFrame of the url to show in notebook with height ht"""
    from IPython.display import IFrame 
    return IFrame(url, width='95%', height=ht) 
showURL('https://en.wikipedia.org/wiki/Bash_(Unix_shell)',400)

## 1. Dropping into BASH (Unix Shell)

Using `%%sh` in a code cell we can access the BASH (Unix Shell) command prompt.

Let us `pwd` or print working directory.

In [2]:
%%sh
pwd

/Users/avelin/git/1MS041/master/jp


In [3]:
%%sh
# this is a comment in BASH shell as it is preceeded by '#'
ls # list the contents of this working directory

00.ipynb
01.ipynb
02.ipynb
03.ipynb
04.ipynb
05.ipynb
06.ipynb
07.ipynb
08.ipynb
09.ipynb
10.ipynb
11.ipynb
12.ipynb
13.ipynb
EXAM2019.ipynb
README.md
Untitled.ipynb
data
images
myHist.png


In [4]:
%%sh
mkdir mydir

In [5]:
%%sh
cd mydir
pwd
ls -al

/Users/avelin/git/1MS041/master/jp/mydir
total 0
drwxr-xr-x  2 sage sage  64 Aug 17 11:40 .
drwxr-xr-x 24 sage sage 768 Aug 17 11:40 ..


In [6]:
%%sh
pwd

/Users/avelin/git/1MS041/master/jp


## "Use the source" by `man`-ning the unknown `command`

By evaluating the next cell, you are using thr `man`ual pages to find more about the command `ls`. You can learn more about any command called `command` by typing `man command` in the BASH shell.

The output of the next cell with command `man ls` will look something like the following:

```
LS(1)                            User Commands                           LS(1)

NAME
       ls - list directory contents

SYNOPSIS
       ls [OPTION]... [FILE]...

DESCRIPTION
       List  information  about  the FILEs (the current directory by default).
       Sort entries alphabetically if none of -cftuvSUX nor --sort  is  speci‐
       fied.

       Mandatory  arguments  to  long  options are mandatory for short options
       too.

       -a, --all
              do not ignore entries starting with .

       -A, --almost-all
              do not list implied . and ..
...
...
...
   Exit status:
       0      if OK,

       1      if minor problems (e.g., cannot access subdirectory),

       2      if serious trouble (e.g., cannot access command-line argument).

AUTHOR
       Written by Richard M. Stallman and David MacKenzie.

REPORTING BUGS
       GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
       Report ls translation bugs to <http://translationproject.org/team/>

COPYRIGHT
       Copyright © 2017 Free Software Foundation, Inc.   License  GPLv3+:  GNU
       GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
       This  is  free  software:  you  are free to change and redistribute it.
       There is NO WARRANTY, to the extent permitted by law.

SEE ALSO
       Full documentation at: <http://www.gnu.org/software/coreutils/ls>
       or available locally via: info '(coreutils) ls invocation'

GNU coreutils 8.28               January 2018                            LS(1)
```

In [7]:
%%sh
man ls

LS(1)                            User Commands                           LS(1)

NAME
       ls - list directory contents

SYNOPSIS
       ls [OPTION]... [FILE]...

DESCRIPTION
       List  information  about  the FILEs (the current directory by default).
       Sort entries alphabetically if none of -cftuvSUX nor --sort  is  speci‐
       fied.

       Mandatory  arguments  to  long  options are mandatory for short options
       too.

       -a, --all
              do not ignore entries starting with .

       -A, --almost-all
              do not list implied . and ..

       --author
              with -l, print the author of each file

       -b, --escape
              print C-style escapes for nongraphic characters

       --block-size=SIZE
              scale sizes by SIZE before printing them; e.g., '--block-size=M'
              prints sizes in units of 1,048,576 bytes; see SIZE format below

       -B, --ignore-backups
              do not list implied entries ending with ~

 

## 2. Grabbing files from internet using curl

In [8]:
%%sh
cd mydir
curl -O http://lamastex.org/datasets/public/SOU/sou/20170228.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29323  100 29323    0     0  22872      0  0:00:01  0:00:01 --:--:-- 22855


In [9]:
%%sh
ls mydir/

20170228.txt


In [10]:
%%sh
cd mydir/
head 20170228.txt 

Donald J. Trump 

February 28, 2017 
Thank you very much. Mr. Speaker, Mr. Vice President, members of Congress, the first lady of the United States ... 
... and citizens of America, tonight, as we mark the conclusion of our celebration of Black History Month, we are reminded of our nation's path toward civil rights and the work that still remains to be done. 
Recent threats ... 
Recent threats targeting Jewish community centers and vandalism of Jewish cemeteries, as well as last week's shooting in Kansas City, remind us that while we may be a nation divided on policies, we are a country that stands united in condemning hate and evil in all of its very ugly forms. 
Each American generation passes the torch of truth, liberty and justice, in an unbroken chain all the way down to the present. That torch is now in our hands. And we will use it to light up the world. 
I am here tonight to deliver a message of unity and strength, and it is a message deeply delivered from my heart. A new chapt

## To have more fun with all SOU addresses
Do the following:


In [11]:
%%sh
mkdir -p mydir # first create a directory called 'mydir'
cd mydir # change into this mydir directory
rm -f sou.tar.gz # remove any file in mydir called sou.tar.gz
curl -O http://lamastex.org/datasets/public/SOU/sou.tar.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3566k  100 3566k    0     0   333k      0  0:00:10  0:00:10 --:--:--  285k


In [12]:
%%sh
pwd
ls -lh mydir

/Users/avelin/git/1MS041/master/jp
total 4.1M
-rw-r--r-- 1 sage sage  29K Aug 17 11:40 20170228.txt
-rw-r--r-- 1 sage sage 3.5M Aug 17 11:40 sou.tar.gz


In [13]:
%%sh
cd mydir 
tar zxvf sou.tar.gz

sou/
sou/18111105.txt
sou/20040120.txt
sou/19061203.txt
sou/18411207.txt
sou/19091207.txt
sou/18701205.txt
sou/19410106.txt
sou/18571208.txt
sou/18891203.txt
sou/18341201.txt
sou/19660112.txt
sou/17981208.txt
sou/19610130.txt
sou/18140920.txt
sou/18011208.txt
sou/18811206.txt
sou/18281202.txt
sou/19840125.txt
sou/18611203.txt
sou/18731201.txt
sou/19400103.txt
sou/19630114.txt
sou/19281204.txt
sou/19221208.txt
sou/19031207.txt
sou/18681209.txt
sou/18431206.txt
sou/18861206.txt
sou/19261207.txt
sou/19271206.txt
sou/19141208.txt
sou/18791201.txt
sou/19131202.txt
sou/19041206.txt
sou/18001111.txt
sou/18041108.txt
sou/20010227.txt
sou/18621201.txt
sou/19251208.txt
sou/19700122.txt
sou/19790125.txt
sou/19870127.txt
sou/20050202.txt
sou/18331203.txt
sou/17961207.txt
sou/18021215.txt
sou/18771203.txt
sou/19890209.txt
sou/18301206.txt
sou/18121104.txt
sou/19580109.txt
sou/20110125.txt
sou/19450106.txt
sou/18031017.txt
sou/19301202.txt
sou/18661203.txt
sou/19520109.txt
sou/19620111.txt
sou/18531

After running the above two cells, you should have all the SOU (State of Union) addresses. By evaluating the next cell's `ls ...` command you should see the SOU files like the following:

```
total 11M
-rw------- 1 raazesh raazesh 6.6K Feb 18  2016 17900108.txt
-rw------- 1 raazesh raazesh 8.3K Feb 18  2016 17901208.txt
-rw------- 1 raazesh raazesh  14K Feb 18  2016 17911025.txt
...
...
...
-rw------- 1 raazesh raazesh  39K Feb 18  2016 20140128.txt
-rw------- 1 raazesh raazesh  38K Feb 18  2016 20150120.txt
-rw------- 1 raazesh raazesh  31K Feb 18  2016 20160112.txt
```

In [14]:
%%sh
ls -lh mydir/sou

total 11M
-rw------- 1 sage sage 6.6K Feb 18  2016 17900108.txt
-rw------- 1 sage sage 8.3K Feb 18  2016 17901208.txt
-rw------- 1 sage sage  14K Feb 18  2016 17911025.txt
-rw------- 1 sage sage  13K Feb 18  2016 17921106.txt
-rw------- 1 sage sage  12K Feb 18  2016 17931203.txt
-rw------- 1 sage sage  18K Feb 18  2016 17941119.txt
-rw------- 1 sage sage  13K Feb 18  2016 17951208.txt
-rw------- 1 sage sage  17K Feb 18  2016 17961207.txt
-rw------- 1 sage sage  13K Feb 18  2016 17971122.txt
-rw------- 1 sage sage  14K Feb 18  2016 17981208.txt
-rw------- 1 sage sage 9.1K Feb 18  2016 17991203.txt
-rw------- 1 sage sage 8.2K Feb 18  2016 18001111.txt
-rw------- 1 sage sage  19K Feb 18  2016 18011208.txt
-rw------- 1 sage sage  13K Feb 18  2016 18021215.txt
-rw------- 1 sage sage  14K Feb 18  2016 18031017.txt
-rw------- 1 sage sage  13K Feb 18  2016 18041108.txt
-rw------- 1 sage sage  17K Feb 18  2016 18051203.txt
-rw------- 1 sage sage  17K Feb 18  2016 18061202.txt
-rw------- 1 sage 

In [15]:
%%sh
head mydir/sou/17900108.txt

George Washington 

January 8, 1790 
Fellow-Citizens of the Senate and House of Representatives: 
I embrace with great satisfaction the opportunity which now presents itself of congratulating you on the present favorable prospects of our public affairs. The recent accession of the important state of North Carolina to the Constitution of the United States (of which official information has been received), the rising credit and respectability of our country, the general and increasing good will toward the government of the Union, and the concord, peace, and plenty with which we are blessed are circumstances auspicious in an eminent degree to our national prosperity. 
In resuming your consultations for the general good you can not but derive encouragement from the reflection that the measures of the last session have been as satisfactory to your constituents as the novelty and difficulty of the work allowed you to hope. Still further to realize their expectations and to secure the blessin

In [16]:
%%sh
head mydir/sou/20160112.txt

Barack Obama 

January 12, 2016 
Mr. Speaker, Mr. Vice President, Members of Congress, my fellow Americans: 
Tonight marks the eighth year I've come here to report on the State of the Union. And for this final one, I'm going to try to make it shorter. I know some of you are antsy to get back to Iowa. 
I also understand that because it's an election season, expectations for what we'll achieve this year are low. Still, Mr. Speaker, I appreciate the constructive approach you and the other leaders took at the end of last year to pass a budget and make tax cuts permanent for working families. So I hope we can work together this year on bipartisan priorities like criminal justice reform, and helping people who are battling prescription drug abuse. We just might surprise the cynics again. 
But tonight, I want to go easy on the traditional list of proposals for the year ahead. Don't worry, I've got plenty, from helping students learn to write computer code to personalizing medical treatments f

An interesting analysis of the textual content of the *State of the Union (SoU)* addresses by all US presidents was done in:

-   [Alix Rule, Jean-Philippe Cointet, and Peter S. Bearman, Lexical shifts, substantive changes, and continuity in State of the Union discourse, 1790–2014, PNAS 2015 112 (35) 10837-10844; doi:10.1073/pnas.1512221112](http://www.pnas.org/content/112/35/10837.full).

![](images/F5.large.png)

[Image source](http://www.pnas.org/content/112/35/10837/F5.large.jpg)

[Fig. 5](http://www.pnas.org/content/112/35/10837.full). A river network captures the flow across history of US political discourse, as perceived by contemporaries. Time moves along the x axis. Clusters on semantic networks of 300 most frequent terms for each of 10 historical periods are displayed as vertical bars. Relations between clusters of adjacent periods are indexed by gray flows, whose density reflects their degree of connection. Streams that connect at any point in history may be considered to be part of the same system, indicated with a single color.

**You** *will be able to carry out such analyses and/or critically reflect on the mathematical statistical assumptions made in such analyses, as you learn more during your programme of study after successfully completing this course.*

## How the `sou.tgz` file was created?

If you are curious, read: [http://lamastex.org/datasets/public/SOU/README.md](http://lamastex.org/datasets/public/SOU/README.md).

Briefly, this is how a website with SOU was scraped by Paul Brouwers and adapted by Raazesh Sainudiin. A data scientist, and more generally a researcher interested in making statistical inference from data that is readily available online in a particular format, is expected to be comfortable with such *web-scraping tasks* (which can be done in more gracious and robust ways using specialised Python libraries). Such tasks also known as *Extract-Load-Transform (ELT)* operations are often time-consuming, expensive andnthe necessary first step towards statistical inference.

### A bit of bash and lynx to achieve the scraping of the state of the union addresses of the US Presidents,
#### by Paul Brouwers

The code below is mainly there to show how the text content of each state of the union address was scraped from the following URL:

* [http://stateoftheunion.onetwothree.net/texts/index.html](http://stateoftheunion.onetwothree.net/texts/index.html)

Such data acquisition tasks is usually the first and cucial step in a data scientist's workflow.

We have done this and put the data in the distributed file system for easy loading into our notebooks for further analysis.  This keeps us from having to install unix programs like `lynx`, `sed`, etc. that are needed in the shell script below.


```
for i in $(lynx --dump http://stateoftheunion.onetwothree.net/texts/index.html | grep texts | grep -v index | sed 's/.*http/http/') ; do lynx --dump $i | tail -n+13 | head -n-14 | sed 's/^\s\+//' | sed -e ':a;N;$!ba;s/\(.\)\n/\1 /g' -e 's/\n/\n\n/' > $(echo $i | sed 's/.*\([0-9]\{8\}\).*/\1/').txt ; done
```

Or in a more atomic form:

```
for i in $(lynx --dump http://stateoftheunion.onetwothree.net/texts/index.html \

        | grep texts \

        | grep -v index \

        | sed 's/.*http/http/')

do 

        lynx --dump $i \

               | tail -n+13 \

               | head -n-14 \

               | sed 's/^\s\+//' \

               | sed -e ':a;N;$!ba;s/\(.\)\n/\1 /g' -e 's/\n/\n\n/' \

               > $(echo $i | sed 's/.*\([0-9]\{8\}\).*/\1/').txt

done
```